src/mlpack/core/dists/discrete_distribution.cpp

```diff
@@ -44,67 +44,97 @@ arma::vec DiscreteDistribution::Random() const
-  probabilities.zeros();
+  // Deal with the difference between vector and martix
+  const arma::mat transObservations = observations.t();
 
-  // Add the probability of each observation.  The addition of 0.5 to the
-  // observation is to turn the default flooring operation of the size_t cast
-  // into a rounding operation.
-  for (size_t i = 0; i < observations.n_cols; i++)
+  // Make sure the observations have same dimension with the probabilities
+  if(transObservations.n_cols != probabilities.size())
   {
-    const size_t obs = size_t(observations(0, i) + 0.5);
-
-    // Ensure that the observation is within the bounds.
-    if (obs >= probabilities.n_elem)
+    Log::Debug << "the obversation must has the same dimension with the probabilities"
+        << "the observation's dimension is" << transObservations.n_cols << "but the dimension of "
+        << "probabilities is" << probabilities.size() << std::endl;
+  }
+  // Get the dimension size of the distribution
+  const size_t dimensions = probabilities.size();
+  // Iterate all the probabilities in each dimension
+  for (size_t i=0; i < dimensions; i++)
+  {
+    // Clear the old probabilities
+    probabilities[i].zeros();
+    for (size_t r=0; r < transObservations.n_rows; r++)
     {
-      Log::Debug << "DiscreteDistribution::Train(): observation " << i
-          << " (" << obs << ") is invalid; observation must be in [0, "
-          << probabilities.n_elem << "] for this distribution." << std::endl;
+      // Add the probability of each observation.  The addition of 0.5 to the
+      // observation is to turn the default flooring operation of the size_t cast
+      // into a rounding observation.
+      const size_t obs = size_t(transObservations(r, i) + 0.5);
+
+      // Ensure that the observation is within the bounds.
+      if (obs >= probabilities[i].n_elem)
+      {
+        Log::Debug << "DiscreteDistribution::Train(): observation " << i
+            << " (" << obs << ") is invalid; observation must be in [0, "
+            << probabilities[i].n_elem << "] for this distribution." << std::endl;       
+      }
+      probabilities[i][obs]++;
     }
 
-    probabilities(obs)++;
+    // Now normailze the distribution.
+    double sum = accu(probabilities[i]);
+    if (sum > 0)
+      probabilities[i] /= sum;
+    else
+      probabilities[i].fill(1.0 / probabilities[i].n_elem); // Force normalization.
   }
-
-  // Now normalize the distribution.
-  double sum = accu(probabilities);
-  if (sum > 0)
-    probabilities /= sum;
-  else
-    probabilities.fill(1 / probabilities.n_elem); // Force normalization.
 }
 
 /**
  * Estimate the probability distribution from the given observations when also
  * given probabilities that each observation is from this distribution.
  */
 void DiscreteDistribution::Train(const arma::mat& observations,
-                                    const arma::vec& probObs)
+                                    const arma::mat& probObs)
```

Comment:
Ah, yeah, I miswrote here.  You should be adding the probability of the observation to `probabilities[i][obs]`.  That means that really it should be

```
probabilities[i][obs] += probObs[r];
```

Sorry about the confusion.